lib/src/common/cache/cache.dart

```diff
@@ -294,6 +301,66 @@ class Cache<TIdentifier, TValue> extends Object with Disposable {
     return new Future.value();
   }
 
+  /// Provides access to a value stored in the cache without informing the
+  /// [CachingStrategy] of a get.
+  ///
+  /// [weakGet] can be thought of sort of like a weak reference when working with
+  /// a garbage collector. [weakGet] won't prevent removal of a [TValue]
+  /// [TIdentifier] pair from the [Cache], [get] or [getAsync] will. In other
+  /// words, it's a get that behaves like `Map`'s does rather than the way
+  /// [Cache]'s does. This can be convenient, for example, when mutating an
+  /// object that is known to be in the cache.
+  ///
+  /// This ideally shouldn't be used to create a reference with a lifetime that
+  /// will exceed the time the item exists in the cache. For example this would be
+  /// a bad consumption pattern:
+  ///
+  /// The delineation to the right of the code indicates when an item will be in
+  /// the cache and when a reference created with [weakGet] would be alive.
+  ///
+  ///     var cache = new Cache<String, Disposable>(new LeastRecentlyUsedStrategy(0));
+  ///     cache.didRemove.listen((CacheContext context) => context.value.dispose());
+  ///     cache.get('thing', () => new Disposable());                                  ┓
+  ///                                                                                  ┃
+  ///     var danglingReference = await cache.weakGet('thing');                        ┃┓
+  ///                                                                                  ┃┃
+  ///     await cache.release('thing');                                                ┛┃
+  ///     // Await a random future to ensure that the callback registered on didRemove  ┃
+  ///     // is called.                                                                 ┃
+  ///     await new Future((){});                                                       ┃
+  ///                                                                                   ┃
+  ///     // This will throw because the instance associated with 'thing' has been      ┃
+  ///     // removed from the cache and disposed on the didRemove callback.             ┃
```

Comment:
Isn't this a problem with the consumption pattern of disposing on removal, and not with the caching class itself?

It seems like a better approach would be to not dispose the object automatically if there are cases where you might access it after it gets removed. Or, alternatively, keep that automatic disposal in place and check whether the object is disposed before interacting with it.

Reply:
Is this the use-case where you'd like to have the behavior, or just an example? I'd like to get more insight into the problem you're trying to solve so I can understand the need for this behavior better.

Reply:
This is intended to be an example of a bad consumption pattern that should be avoided. 

### For a little bit of clarification on the problem trying to be solved:

In TDC we have a least-recently-used loaded experience cache. When an experience goes on screen it's loaded and put into the cache. When it goes off screen it's marked as `released` meaning that's it's eligible for being unloaded (which will happen when it is removed from the cache). The idea is to prevent the user having to wait for load every time an experience is briefly scrolled off screen then back on while still ensuring that the experience will be removed from memory at some point. 

There are several cases where something needs to call a method on one of these loaded experiences.  This was done by doing a cache `get` for the experience then calling the method on the returned experience. Calling one of these methods implies that you expect the experience to be loaded and onscreen, if you're calling one of these methods on something that is potentially unloaded and off screen it's an error. Some errant go code was calling one of these methods at an improper time (after the same code has informed TDC that it's ok for the experience to be unloaded i.e `release`, but not long enough for the experience to be unloaded). The cache `get` required to get at the experience removed it from the pool of experiences marked as eligible for removal, meaning that the experience would never unloaded.

This API is an attempt to create a way  get an item from the cache, that is expected to be there, to mutate it or access some property that won't lead to retaining it in the cache. The errant code calling a method at the wrong time also needs to be fixed, but we believe it would be a more robust consumption pattern to have calls that come at the wrong time to get a `null` rather than leading to unintentional retention.

### As for the consumption pattern: 
Usually you wouldn't hide an object behind a cache unless it's pretty heavy, so I would think that disposing on removal would be a pretty usual consumption pattern.